libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Analytics/SFSDKAILTNPublisher.m

```diff
@@ -53,41 +51,32 @@ + (BOOL) publish:(NSArray *) events {
-        NSMutableURLRequest *request = [NSMutableURLRequest requestWithURL:loggingEndpointUrl
-                                                               cachePolicy:NSURLRequestReloadIgnoringCacheData
-                                                           timeoutInterval:60.0];
-        [request setHTTPMethod:@"POST"];
-        [request setValue:userAgent forHTTPHeaderField:@"User-Agent"];
-        [request setValue:@"false" forHTTPHeaderField:@"X-Chatter-Entity-Encoding"];
-        [request setValue:token forHTTPHeaderField:@"Authorization"];
-        [request setValue:@"application/json" forHTTPHeaderField:@"Content-Type"];
+    NSString *path = [NSString stringWithFormat:@"/%@/%@", kSFRestDefaultAPIVersion, kRestApiSuffix];
+    SFRestRequest *request = [SFRestRequest requestWithMethod:SFRestMethodPOST path:path queryParams:nil];
 
-        // Adds GZIP compression.
-        NSString *bodyString = [[self class] dictionaryAsJSONString:bodyDictionary];
-        NSData *bodyData = [bodyString dataUsingEncoding:NSUTF8StringEncoding];
-        NSData *postData = [bodyData gzipDeflate];
-        [request setValue:@"gzip" forHTTPHeaderField:@"Content-Encoding"];
-        [request setValue:[NSString stringWithFormat:@"%lu", (unsigned long)[postData length]] forHTTPHeaderField:@"Content-Length"];
-        [request setHTTPBody:postData];
-        NSHTTPURLResponse* response = nil;
-        [NSURLConnection sendSynchronousRequest:request returningResponse:&response error:error];
-        return response;
-    };
-    NSHTTPURLResponse *response = makeSynchronousRequest(&error);
-    NSInteger code = [response statusCode];
-    if (error) {
-        [SFLogger log:[self class] level:SFLogLevelError format:@"Upload failed %ld %@", (long)[error code], [error localizedDescription]];
-        return NO;
-    } else if (code >= 200 && code < 300) {
-        return YES;
+    // Adds GZIP compression.
+    NSString *bodyString = [[self class] dictionaryAsJSONString:bodyDictionary];
+    NSData *bodyData = [bodyString dataUsingEncoding:NSUTF8StringEncoding];
+    NSData *postData = [bodyData gzipDeflate];
+    [request setCustomRequestBodyData:postData contentType:@"application/json"];
+    [request setHeaderValue:@"gzip" forHeaderName:@"Content-Encoding"];
+    [request setHeaderValue:[NSString stringWithFormat:@"%lu", (unsigned long)[postData length]] forHeaderName:@"Content-Length"];
+    __block BOOL finished = NO;
+    __block BOOL success = NO;
+    [[SFRestAPI sharedInstance] sendRESTRequest:request failBlock:^(NSError *e) {
+        finished = YES;
+            if (e) {
+                [SFLogger log:[self class] level:SFLogLevelError format:@"Upload failed %ld %@", (long)[e code], [e localizedDescription]];
+            }
+        } completeBlock:^(id response) {
+            finished = YES;
+            success = YES;
+        }
+    ];
+    while(!finished) {
```

Comment:
An asynchronous pattern here adds complexity to the code and some uncertainty in the order of requests/responses. The API contract of the publisher interface requires `publish` to return `BOOL` indicating success of the request. This allows the library to then queue the next publisher and use this status to determine whether to delete stored records or not. If we were to change this to an async pattern using completion blocks, it would make the caller's code a little more messy. For example, do we sequentially line up publishers or fire them all at the same time? Also, when the completion blocks are triggered, which publisher is it coming from? Also, the logic to remove records requires that all publishers were successful. This requires more logic in the completion blocks and counting how many publishers came back with a status (especially if we trigger them parallely). We would be pushing the sync portion one level lower to the `publish` method in the manager and adding code complexity. Hence, I went with the sync approach here, for this special case.